portfolio/src/main/java/com/google/sps/servlets/CommentServlet.java

```diff
@@ -86,21 +87,47 @@ private Comment entityToComment(Entity entity) {
     String email = (String) entity.getProperty("email");
     Date timestamp = (Date) entity.getProperty("timestamp");
     String message = (String) entity.getProperty("message");
+    float sentimentScore;
+    try {
+      sentimentScore = ((Number) entity.getProperty("sentimentScore")).floatValue();
+    } catch(NullPointerException e) {
+      sentimentScore = 0f;
+    }
     
-    Comment comment = new Comment(author, email, timestamp, message);
+    Comment comment = 
+        new Comment(author, email, timestamp, message, sentimentScore);
     return comment;
   }
 
   private Entity commentRequestToEntity(HttpServletRequest request) {
     String author = request.getParameter("name");
     Date timestamp = new Date();
     String message = request.getParameter("message");
+    float sentimentScore;
+    try {
+      sentimentScore = analyzeSentiment(message);
+    } catch(IOException e) {
+      sentimentScore = 0f;
+    }
 
     Entity commentEntity = new Entity(entityKind, homepageCommentKey);
     commentEntity.setProperty("author", author);
     commentEntity.setProperty("email", userEmail);
     commentEntity.setProperty("timestamp", timestamp);
     commentEntity.setProperty("message", message);
+    commentEntity.setProperty("sentimentScore", sentimentScore);
+    
     return commentEntity;
   }
+
+  private float analyzeSentiment(String message) throws IOException {
+    Document document = 
+        Document.newBuilder().setContent(message).setType(Document.Type.PLAIN_TEXT).build();
+    LanguageServiceClient languageService = LanguageServiceClient.create();
+    Sentiment sentiment = languageService.analyzeSentiment(document).getDocumentSentiment();
+    float magnitude = sentiment.getMagnitude();
+    float score = sentiment.getScore();
+    languageService.close();
```

Comment:
in this particular case, it seems like only languageServiceClient.create() throws an exception, so this might be okay. In general, it might be safer to use the "try-with-resources" construct in java. https://googleapis.dev/java/google-cloud-language/latest/com/google/cloud/language/v1beta2/LanguageServiceClient.html

This is so that if an exception happens before you close the service, you don't accidentally leak it.